### PR TITLE
[exodus_] Add plugin to monitore number of applications, reports and trackers

### PR DESCRIPTION
**Pull request by codeurimpulsif**
## Goal

This plugin monitore number of applications, reports and trackers of an [Exodus](https://github.com/Exodus-Privacy/exodus/) instance.

## Informations

This plugin is based on [the Exodus API documentation](https://github.com/Exodus-Privacy/exodus/blob/v1/doc/api.md) and can be used with any Exodus instance starting from [version 1.25.0](https://github.com/Exodus-Privacy/exodus/releases/tag/v1.25.0).

### CODE REVIEW

File: plugins/exodus/exodus_ [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    if count.status_code == 200:
+        print(type + '.value', count.json()['count'])
+
+
+def main():
+    try:
+        mode = sys.argv[1]
+    except IndexError:
+        mode = ""
+    wildcard = sys.argv[0].split("exodus_")[1].split("_")[0]
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:74]: What is the purpose of the final `split`?

File: plugins/exodus/exodus_ [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    if mode == "suggest":
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:76]: You should `exit` after emitting the `suggest` items.

File: plugins/exodus/exodus_ [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+=cut
+"""
+
+import os
+import sys
+import requests
+
+
+def print_count(type):
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:61]: You should avoid overriding names of `builtins` like `type`.

File: plugins/exodus/exodus_ [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+ #%# family=manual
+ #%# capabilities=suggest
+
+=cut
+"""
+
+import os
+import sys
+import requests
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:58]: `request` is not part of the standard library.
You could replace it with `urllib` (part of the standard library), e.g.
```python
request = urllib.request.Request(endpoint, method="POST", headers=headers)
with urllib.request.urlopen(request) as connection:
    count = json.loads(connection.read())["count"]
```
Or you should use `try/except` for the import and emit an error message to `stderr`, if the module is missing.
It depends on how kind you want to be towards the users :)

File: plugins/exodus/exodus_ [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    ln -s /usr/share/munin/plugins/exodus_ /etc/munin/plugins/exodus_reports
+    ln -s /usr/share/munin/plugins/exodus_ /etc/munin/plugins/exodus_trackers
+
+=back
+
+After the installation you need to restart your munin-node:
+
+=over 2
+
+    systemctl restart munin-node
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:27]: I would recommend to omit this systemd-specific command. A generic hint  (e.g. "restart the munin-node service") should be sufficient and works everywhere - now and in the future.

File: plugins/exodus/exodus_ [hunk 6]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
+"""
+=head1 NAME
+
+exodus_ - Exodus wildcard-plugin to monitor an L<Exodus|https://github.com/Exodus-Privacy/exodus/>
+instance.
+
+This wildcard plugin provides at the moment only the suffixes C<applications>, C<reports>
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:8]: Maybe remove `only` here? Or state, which other parts you consider to be missing.

File: plugins/exodus/exodus_ [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+=over 2
+
+    [exodus_*]
+    env.exodus_url https://reports.exodus-privacy.eu.org
+    env.exodus_token your-api-token
+
+=back
+
+=head1 AUTHORS
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:44]: Please add a *License* section, thanks!

File: plugins/exodus/exodus_ [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        else:
+            print_count('reports')
+    elif wildcard == "trackers":
+        if mode == "config":
+            print("graph_title Exodus trackers")
+            print("graph_vlabel trackers")
+            print("graph_category exodus")
+            print("trackers.label Trackers")
+        else:
+            print_count('trackers')
```
Comment (sumpfralle) [inline plugins/exodus/exodus_:104]: Maybe run `black` in order to unify the quoting style (single or double quotes) in the script?

**Review by sumpfralle — COMMENTED** [review]
Thank you for contributing this new plugin!

Regarding the name: do you think, that the plugin should cover the whole `exodus_` namespace? Or maybe something like `exodus_api_` would be more specific?
(I do not know the exodus application, thus I cannot compare the scope of your plugin with the full scope of exodus)

Please take a look at my comments. Most are just style hints. Thank you!

File: plugins/exodus/exodus_ (hunk 1 above)
Comment (codeurimpulsif) [reply plugins/exodus/exodus_:74]: my bad, it was from another plugin for another use, I remove it

**Comment by codeurimpulsif** [comment]
> Regarding the name: do you think, that the plugin should cover the whole `exodus_` namespace? Or maybe something like `exodus_api_` would be more specific? (I do not know the exodus application, thus I cannot compare the scope of your plugin with the full scope of exodus)

Yes, I think it should cover the whole namespace (I talked about that with one of the main developers of Exodus)

**Comment by sumpfralle** [comment]
Thank you for going through my nitpicks and for contributing this new plugin!

**Pull request merged**